internal/internal.go

```diff
@@ -190,12 +190,25 @@ func (p *envoyExtAuthzGrpcServer) listen() {
 }
 
 func (p *envoyExtAuthzGrpcServer) Check(ctx context.Context, req *ext_authz.CheckRequest) (resp *ext_authz.CheckResponse, err error) {
+
+	// Check if context is cancelled or deadline is reached.
+	exit := make(chan struct{})
+	defer close(exit)
+	go waitForDone(ctx, exit, func() {
+		err = fmt.Errorf("context deadline reached due to server overload")
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can we check for `ctx.Err()` rather than using the done channel directly? If it returns anything except nil we know that done is complete.

Reply:
Updated

Reply:
Ah sry, i guess i meant not using the background thread and instead doing something more simple like:

```
if ctx.Err() != nil {
	return nil, errors.Wrap(ctx.Err(), "context deadline reached due to server overload")
}
```

Was the intention to have that background thread waiting to catch cancellations during evaluation too?

Reply:
You're right. We don't need to do that.